### ISSUE TITLE
Improve FIX message handling

### ISSUE THREAD

**Issue by jvirtanen** [paritytrading/philadelphia#95]
Currently Philadelphia represents both incoming and outgoing FIX messages as instances of the [`FIXMessage`](https://github.com/paritytrading/philadelphia/blob/2.0.0/libraries/core/src/main/java/com/paritytrading/philadelphia/FIXMessage.java) class. A `FIXMessage` object is a sequential collection of tag-value pairs. Each tag is represented by an `int`, and each value is represented by a  [`FIXValue`](). `FIXValue` provides both read and write access to a value.

Internally, each `FIXValue` maintains an independent byte array as a backing store. On each read, the FIX value representation is parsed into an appropriate Java type, such as `int` or a  `double`. Similarly, on each write, a Java value is formatted into the corresponding FIX value representation. The size of the backing store is defined by [`FIXConfig#getFieldCapacity()`](https://github.com/paritytrading/philadelphia/blob/master/libraries/core/src/main/java/com/paritytrading/philadelphia/FIXConfig.java#L124).

When a message is parsed from a `ByteBuffer`, a `FIXMessage` instance essentially scatters the `ByteBuffer` contents into the `FIXValue` objects it holds. When a message is formatted into a `ByteBuffer`, the opposite takes place.

I suggest the following improvement to this architecture. A `FIXMessage` will maintain one contiguous byte array as a backing store. For outgoing messages, `FIXMessage` will provide methods that append a field of a specific type into the message. For example:
```java
/**
 * Add an integer field.
 *
 * @param tag the tag
 * @param value the value
 */
public void addInt(int tag, int value);
```
Appending a field will format the both the tag and the value directly into the backing store. This way, sending the message will only involve copying the backing store into `ByteBuffer`. (As a subsequent step, the backing store could be implemented directly as `ByteBuffer` to avoid the copy altogether.)

For incoming messages, `FIXMessage` will provide methods for accessing values of specific types. For example:
```java
/**
 * Get the integer value at the specified index.
 *
 * @param index the index
 * @return the integer value
 * @throws FIXValueFormatException if the value is not an integer
 */
public int intAt(int index);

/**
 * Get the integer value of the first instance of a field with the
 * specified tag.
 *
 * @param tag the tag
 * @return the integer value
 * @throws FIXValueFormatException if the value is not an integer
 */
public int intOf(int tag);
```
The existing methods for accessing tags will remain. `FIXValue` will be removed, as all access to values is handled directly through `FIXMessage`. Notably also it will no longer be possible to update values in outgoing messages: all fields must be written out in sequential order. As the  purpose of this change is to improve the performance of the library, this additional limitation feels reasonable to me.

Thoughts?

**Note.** This would be breaking change and therefore scheduled for Philadelphia 2.0.0.

**Comment by khsergvl** [comment]
@jvirtanen I could take care about that, after(as i'm working for that now):

- Add 'expect' command #92  
- Add FIX message parser benchmark #85

**Comment by jvirtanen** [comment]
@khsergvl Like #92, this is currently more of an invitation for discussion on how to best achieve the desired outcome. 🙂 Great to hear that you are looking into #85! 👍
